### PR TITLE
Web - The horizontal line separating the header and the messages

### PR DESCRIPTION
**Pull request by zukilover**
### Details

Changes `useRef` implementation for `shouldHideThreadDividerLine` flag with `useState` to render the horizontal line correctly when a message is marked as unread.

### Fixed Issues

$ [#26529](https://github.com/Expensify/App/issues/26529)
PROPOSAL: [#26529](https://github.com/Expensify/App/issues/26529#issuecomment-1707444942)

### Tests

1. Click on the FAB and request Money from any user this will take you to that chat room
2. Click on the IOU message to open the IOU report room
3. Hover over the first Message in the IOU report room and mark it as unread
4. An unread mark should appear and the grey horizontal line should be hidden

- [x] Verify that no errors appear in the JS console

### Offline tests

Same as Tests

### QA Steps

Same as Tests

- [x] Verify that no errors appear in the JS console

### PR Author Checklist

- [x] I linked the correct issue in the `### Fixed Issues` section above
- [x] I wrote clear testing steps that cover the changes made in this PR
    - [x] I added steps for local testing in the `Tests` section
    - [x] I added steps for the expected offline behavior in the `Offline steps` section
    - [x] I added steps for Staging and/or Production testing in the `QA steps` section
    - [x] I added steps to cover failure scenarios (i.e. verify an input displays the correct error message if the entered data is not correct)
    - [x] I turned off my network connection and tested it while offline to ensure it matches the expected behavior (i.e. verify the default avatar icon is displayed if app is offline)
    - [x] I tested this PR with a [High Traffic account](https://github.com/Expensify/App/blob/main/contributingGuides/CONTRIBUTING.md#high-traffic-accounts) against the staging or production API to ensure there are no regressions (e.g. long loading states that impact usability).
- [x] I included screenshots or videos for tests on [all platforms](https://github.com/Expensify/App/blob/main/contributingGuides/CONTRIBUTING.md#make-sure-you-can-test-on-all-platforms)
- [x] I ran the tests on **all platforms** & verified they passed on:
    - [x] Android / native
    - [x] Android / Chrome
    - [x] iOS / native
    - [x] iOS / Safari
    - [x] MacOS / Chrome / Safari
    - [x] MacOS / Desktop
- [x] I verified there are no console errors (if there's a console error not related to the PR, report it or open an issue for it to be fixed)
- [x] I followed proper code patterns (see [Reviewing the code](https://github.com/Expensify/App/blob/main/contributingGuides/PR_REVIEW_GUIDELINES.md#reviewing-the-code))
    - [x] I verified that any callback methods that were added or modified are named for what the method does and never what callback they handle (i.e. `toggleReport` and not `onIconClick`)
    - [x] I verified that the left part of a conditional rendering a React component is a boolean and NOT a string, e.g. `myBool && <MyComponent />`.
    - [x] I verified that comments were added to code that is not self explanatory
    - [x] I verified that any new or modified comments were clear, correct English, and explained "why" the code was doing something instead of only explaining "what" the code was doing.
    - [x] I verified any copy / text shown in the product is localized by adding it to `src/languages/*` files and using the [translation method](https://github.com/Expensify/App/blob/4bd99402cebdf4d7394e0d1f260879ea238197eb/src/components/withLocalize.js#L60)
      - [x] If any non-english text was added/modified, I verified the translation was requested/reviewed in #expensify-open-source and it was approved by an internal Expensify engineer. Link to Slack message:
    - [x] I verified all numbers, amounts, dates and phone numbers shown in the product are using the [localization methods](https://github.com/Expensify/App/blob/4bd99402cebdf4d7394e0d1f260879ea238197eb/src/components/withLocalize.js#L60-L68)
    - [x] I verified any copy / text that was added to the app is grammatically correct in English. It adheres to proper capitalization guidelines (note: only the first word of header/labels should be capitalized), and is approved by marketing by adding the `Waiting for Copy` label for a copy review on the original GH to get the correct copy.
    - [x] I verified proper file naming conventions were followed for any new files or renamed files. All non-platform specific files are named after what they export and are not named "index.js". All platform-specific files are named for the platform the code supports as outlined in the README.
    - [x] I verified the JSDocs style guidelines (in [`STYLE.md`](https://github.com/Expensify/App/blob/main/contributingGuides/STYLE.md#jsdocs)) were followed
- [x] If a new code pattern is added I verified it was agreed to be used by multiple Expensify engineers
- [x] I followed the guidelines as stated in the [Review Guidelines](https://github.com/Expensify/App/blob/main/contributingGuides/PR_REVIEW_GUIDELINES.md)
- [x] I tested other components that can be impacted by my changes (i.e. if the PR modifies a shared library or component like `Avatar`, I verified the components using `Avatar` are working as expected)
- [x] I verified all code is DRY (the PR doesn't include any logic written more than once, with the exception of tests)
- [x] I verified any variables that can be defined as constants (ie. in CONST.js or at the top of the file that uses the constant) are defined as such
- [x] I verified that if a function's arguments changed that all usages have also been updated correctly
- [x] If a new component is created I verified that:
    - [x] A similar component doesn't exist in the codebase
    - [x] All props are defined accurately and each prop has a `/** comment above it */`
    - [x] The file is named correctly
    - [x] The component has a clear name that is non-ambiguous and the purpose of the component can be inferred from the name alone
    - [x] The only data being stored in the state is data necessary for rendering and nothing else
    - [x] If we are not using the full Onyx data that we loaded, I've added the proper selector in order to ensure the component only re-renders when the data it is using changes
    - [x] For Class Components, any internal methods passed to components event handlers are bound to `this` properly so there are no scoping issues (i.e. for `onClick={this.submit}` the method `this.submit` should be bound to `this` in the constructor)
    - [x] Any internal methods bound to `this` are necessary to be bound (i.e. avoid `this.submit = this.submit.bind(this);` if `this.submit` is never passed to a component event handler like `onClick`)
    - [x] All JSX used for rendering exists in the render method
    - [x] The component has the minimum amount of code necessary for its purpose, and it is broken down into smaller components in order to separate concerns and functions
- [x] If any new file was added I verified that:
    - [x] The file has a description of what it does and/or why is needed at the top of the file if the code is not self explanatory
- [x] If a new CSS style is added I verified that:
    - [x] A similar style doesn't already exist
    - [x] The style can't be created with an existing [StyleUtils](https://github.com/Expensify/App/blob/main/src/styles/StyleUtils.js) function (i.e. `StyleUtils.getBackgroundAndBorderStyle(themeColors.componentBG)`)
- [x] If the PR modifies code that runs when editing or sending messages, I tested and verified there is no unexpected behavior for all supported markdown - URLs, single line code, code blocks, quotes, headings, bold, strikethrough, and italic.
- [x] If the PR modifies a generic component, I tested and verified that those changes do not break usages of that component in the rest of the App (i.e. if a shared library or component like `Avatar` is modified, I verified that `Avatar` is working as expected in all cases)
- [x] If the PR modifies a component related to any of the existing Storybook stories, I tested and verified all stories for that component are still working as expected.
- [x] If the PR modifies a component or page that can be accessed by a direct deeplink, I verified that the code functions as expected when the deeplink is used - from a logged in and logged out account.
- [x] If a new page is added, I verified it's using the `ScrollView` component to make it scrollable when more elements are added to the page.
- [x] If the `main` branch was merged into this PR after a review, I tested again and verified the outcome was still expected according to the `Test` steps.
- [x] I have checked off every checkbox in the PR author checklist, including those that don't apply to this PR.

### Screenshots/Videos
<details>
<summary>Web</summary>

[web.webm](https://github.com/Expensify/App/assets/371291/e24548f2-7567-40c4-b28e-abf692e50a0d)

</details>

<details>
<summary>Mobile Web - Chrome</summary>

[mweb.chrome.webm](https://github.com/Expensify/App/assets/371291/556ac0f5-d123-49e5-bfe7-823e27a2f61c)

</details>

<details>
<summary>Mobile Web - Safari</summary>

https://github.com/Expensify/App/assets/371291/55cdcc9c-8ada-43e3-b5b4-ff01115092b2

</details>

<details>
<summary>Desktop</summary>

[desktop.webm](https://github.com/Expensify/App/assets/371291/748b603f-275d-43e7-9e9f-98f4ee067658)

</details>

<details>
<summary>iOS</summary>

https://github.com/Expensify/App/assets/371291/e6cbbbbb-1173-4a21-a515-dca10e36667a

</details>

<details>
<summary>Android</summary>

[android.webm](https://github.com/Expensify/App/assets/371291/310f1cd0-b8e8-466f-9431-99bd28fd476f)

</details>

### CODE REVIEW

**Comment by srikarparsi** [comment]
@deetergp, reassigning to you because I believe you were the internal engineer for this [issue](https://github.com/Expensify/App/issues/26529).

**Comment by zukilover** [comment]
> I have read the CLA Document and I hereby sign the CLA

**_I have read the CLA Document and I hereby sign the CLA_**

**Comment by zukilover** [comment]
recheck

**Comment by deetergp** [comment]
Also adding @robertKozik since he was the C+ on [the issue](https://github.com/Expensify/App/issues/26529).

**Comment by robertKozik** [comment]
Thanks @deetergp I'll review it as soon as possible

**Comment by deetergp** [comment]
The code looks good, but in testing, I don't love that it jumps oddly when going from the gray line to the green line. Is there any way we can reign that in?

**Comment by robertKozik** [comment]
@zukilover please check your commits:
1) We have one lint error here - please fix it
2) Your commit is not signed properly - for help please refer to point 10 from [CONTRIBUTING.md](https://github.com/Expensify/App/blob/37feea67f55c429776d0e10be679f1a385eca7da/contributingGuides/CONTRIBUTING.md)

**Comment by zukilover** [comment]
@robertKozik If I sign the next commit, will it solve the gpg sign issue?

**Comment by zukilover** [comment]
@deetergp I will add transition animation for the separator

**Comment by robertKozik** [comment]
> @robertKozik If I sign the next commit, will it solve the gpg sign issue?

I'm afraid not. You'll need to retroactively sign all the commits in the branch. 

But I see you force-pushed the signed one only, It will be sufficient

**Comment by zukilover** [comment]
I have added animated view for the spacer. it should look smoother now @deetergp 

[animated.webm](https://github.com/Expensify/App/assets/371291/ebada72d-440a-4b28-915a-81c7e9a64bfb)

**Comment by zukilover** [comment]
I have fixed the lint issues @robertKozik , please check. Thanks.

**Comment by zukilover** [comment]
Need to force-push for signed gpg.

**Review by robertKozik — CHANGES_REQUESTED** [review]
couple suggestions

**Comment by zukilover** [comment]
I have applied your suggestions @robertKozik. Thanks.

File: src/components/SpacerView.js [hunk 1]
```diff
@@ -0,0 +1,48 @@
...
+     */
+    // eslint-disable-next-line react/forbid-prop-types
+    style: PropTypes.oneOfType([PropTypes.object, PropTypes.array]),
+};
+
+const defaultProps = {
+    style: [],
+};
+
+function SpacerView({shouldShow, style}) {
```
Comment (robertKozik) [inline src/components/SpacerView.js:24]: Please opt for [declaring default values when descructuring](https://github.com/namhihi237/App/blob/9ea79cdaf6b2ef4c49bdafc3cc03ad83fd74dcc7/contributingGuides/STYLE.md#L213-L238)

File: src/components/SpacerView.js [hunk 2]
```diff
@@ -0,0 +1,48 @@
...
+    React.useEffect(() => {
+        const duration = 300;
+        const values = {
+            marginVertical: shouldShow ? 8 : 0,
+            borderBottomWidth: shouldShow ? 1 : 0,
+        };
+        marginVertical.value = withTiming(values.marginVertical, {duration});
+        borderBottomWidth.value = withTiming(values.borderBottomWidth, {duration});
+    }, [shouldShow, borderBottomWidth, marginVertical]);
+
```
Comment (robertKozik) [inline src/components/SpacerView.js:41]: I was referring to hoist the values of duration and margins to the `CONST.ts` file so we won't have any hardcoded 8 1 or 0's inside component

**Review by robertKozik — CHANGES_REQUESTED** [review]

**Comment by zukilover** [comment]
@robertKozik I've refactored them into the CONST file, let me know if that works.

**Comment by zukilover** [comment]
P.S.: I need to leave the `SpacerView.defaultProps = defaultProps` as is, otherwise the linter will complaint:
```
propType "style" is not required, but has no corresponding defaultProps declaration
```

**Comment by robertKozik** [comment]
## Reviewer Checklist

- [X] I have verified the author checklist is complete (all boxes are checked off).
- [X] I verified the correct issue is linked in the `### Fixed Issues` section above
- [X] I verified testing steps are clear and they cover the changes made in this PR
    - [X] I verified the steps for local testing are in the `Tests` section
    - [X] I verified the steps for Staging and/or Production testing are in the `QA steps` section
    - [X] I verified the steps cover any possible failure scenarios (i.e. verify an input displays the correct error message if the entered data is not correct)
    - [X] I turned off my network connection and tested it while offline to ensure it matches the expected behavior (i.e. verify the default avatar icon is displayed if app is offline)
- [X] I checked that screenshots or videos are included for tests on [all platforms](https://github.com/Expensify/App/blob/main/contributingGuides/CONTRIBUTING.md#make-sure-you-can-test-on-all-platforms)
- [X] I included screenshots or videos for tests on [all platforms](https://github.com/Expensify/App/blob/main/contributingGuides/CONTRIBUTING.md#make-sure-you-can-test-on-all-platforms)
- [X] I verified tests pass on **all platforms** & I tested again on:
    - [X] Android / native
    - [X] Android / Chrome
    - [X] iOS / native
    - [X] iOS / Safari
    - [X] MacOS / Chrome / Safari
    - [X] MacOS / Desktop
- [X] If there are any errors in the console that are unrelated to this PR, I either fixed them (preferred) or linked to where I reported them in Slack
- [X] I verified proper code patterns were followed (see [Reviewing the code](https://github.com/Expensify/App/blob/main/contributingGuides/PR_REVIEW_GUIDELINES.md#reviewing-the-code))
    - [X] I verified that any callback methods that were added or modified are named for what the method does and never what callback they handle (i.e. `toggleReport` and not `onIconClick`).
    - [X] I verified that the left part of a conditional rendering a React component is a boolean and NOT a string, e.g. `myBool && <MyComponent />`.
    - [X] I verified that comments were added to code that is not self explanatory
    - [X] I verified that any new or modified comments were clear, correct English, and explained "why" the code was doing something instead of only explaining "what" the code was doing.
    - [X] I verified any copy / text shown in the product is localized by adding it to `src/languages/*` files and using the [translation method](https://github.com/Expensify/App/blob/4bd99402cebdf4d7394e0d1f260879ea238197eb/src/components/withLocalize.js#L60)
    - [X] I verified all numbers, amounts, dates and phone numbers shown in the product are using the [localization methods](https://github.com/Expensify/App/blob/4bd99402cebdf4d7394e0d1f260879ea238197eb/src/components/withLocalize.js#L60-L68)
    - [X] I verified any copy / text that was added to the app is grammatically correct in English. It adheres to proper capitalization guidelines (note: only the first word of header/labels should be capitalized), and is approved by marketing by adding the `Waiting for Copy` label for a copy review on the original GH to get the correct copy.
    - [X] I verified proper file naming conventions were followed for any new files or renamed files. All non-platform specific files are named after what they export and are not named "index.js". All platform-specific files are named for the platform the code supports as outlined in the README.
    - [X] I verified the JSDocs style guidelines (in [`STYLE.md`](https://github.com/Expensify/App/blob/main/contributingGuides/STYLE.md#jsdocs)) were followed
- [X] If a new code pattern is added I verified it was agreed to be used by multiple Expensify engineers
- [X] I verified that this PR follows the guidelines as stated in the [Review Guidelines](https://github.com/Expensify/App/blob/main/contributingGuides/PR_REVIEW_GUIDELINES.md)
- [X] I verified other components that can be impacted by these changes have been tested, and I retested again (i.e. if the PR modifies a shared library or component like `Avatar`, I verified the components using `Avatar` have been tested & I retested again)
- [X] I verified all code is DRY (the PR doesn't include any logic written more than once, with the exception of tests)
- [X] I verified any variables that can be defined as constants (ie. in CONST.js or at the top of the file that uses the constant) are defined as such
- [X] If a new component is created I verified that:
    - [X] A similar component doesn't exist in the codebase
    - [X] All props are defined accurately and each prop has a `/** comment above it */`
    - [X] The file is named correctly
    - [X] The component has a clear name that is non-ambiguous and the purpose of the component can be inferred from the name alone
    - [X] The only data being stored in the state is data necessary for rendering and nothing else
    - [X] For Class Components, any internal methods passed to components event handlers are bound to `this` properly so there are no scoping issues (i.e. for `onClick={this.submit}` the method `this.submit` should be bound to `this` in the constructor)
    - [X] Any internal methods bound to `this` are necessary to be bound (i.e. avoid `this.submit = this.submit.bind(this);` if `this.submit` is never passed to a component event handler like `onClick`)
    - [X] All JSX used for rendering exists in the render method
    - [X] The component has the minimum amount of code necessary for its purpose, and it is broken down into smaller components in order to separate concerns and functions
- [X] If any new file was added I verified that:
    - [X] The file has a description of what it does and/or why is needed at the top of the file if the code is not self explanatory
- [X] If a new CSS style is added I verified that:
    - [X] A similar style doesn't already exist
    - [X] The style can't be created with an existing [StyleUtils](https://github.com/Expensify/App/blob/main/src/styles/StyleUtils.js) function (i.e. `StyleUtils.getBackgroundAndBorderStyle(themeColors.componentBG`)
- [X] If the PR modifies code that runs when editing or sending messages, I tested and verified there is no unexpected behavior for all supported markdown - URLs, single line code, code blocks, quotes, headings, bold, strikethrough, and italic.
- [X] If the PR modifies a generic component, I tested and verified that those changes do not break usages of that component in the rest of the App (i.e. if a shared library or component like `Avatar` is modified, I verified that `Avatar` is working as expected in all cases)
- [X] If the PR modifies a component related to any of the existing Storybook stories, I tested and verified all stories for that component are still working as expected.
- [X] If the PR modifies a component or page that can be accessed by a direct deeplink, I verified that the code functions as expected when the deeplink is used - from a logged in and logged out account.
- [X] If a new page is added, I verified it's using the `ScrollView` component to make it scrollable when more elements are added to the page.
- [X] If the `main` branch was merged into this PR after a review, I tested again and verified the outcome was still expected according to the `Test` steps.
- [X] I have checked off every checkbox in the PR reviewer checklist, including those that don't apply to this PR.

### Screenshots/Videos
<details>
<summary>Web</summary>


https://github.com/Expensify/App/assets/61146594/0575ddb0-cbf1-4c79-a840-a46a8dc935c8



</details>

<details>
<summary>Mobile Web - Chrome</summary>


https://github.com/Expensify/App/assets/61146594/37ae9319-9071-49ae-987a-5efde79d37b8



</details>

<details>
<summary>Mobile Web - Safari</summary>


https://github.com/Expensify/App/assets/61146594/234078f0-e183-45e4-95d7-41edb6292a50



</details>

<details>
<summary>Desktop</summary>


https://github.com/Expensify/App/assets/61146594/6606aa70-e05b-42c3-be8b-b533874afe23



</details>

<details>
<summary>iOS</summary>


https://github.com/Expensify/App/assets/61146594/d52470f5-dde9-4e5d-943b-37dd9f039ca7



</details>

<details>
<summary>Android</summary>


https://github.com/Expensify/App/assets/61146594/2b387500-fecd-4e4f-a7b2-9c4f7c5b7af4



</details>

File: src/components/SpacerView.js [hunk 3]
```diff
@@ -0,0 +1,49 @@
...
+     * Should we show the spacer
+     */
+    shouldShow: PropTypes.bool.isRequired,
+
+    /**
+     * Array of style objects
+     * @default []
+     */
+    // eslint-disable-next-line react/forbid-prop-types
+    style: PropTypes.oneOfType([PropTypes.object, PropTypes.array]),
```
Comment (robertKozik) [inline src/components/SpacerView.js:18]: One last thing - please use already defined style propType inside `stylePropTypes.js`

Comment (zukilover) [reply src/components/SpacerView.js]: Sure, this is updated now. Thanks.

**Comment by deetergp** [comment]
Now when I test this, I don't get a green line at all. Am I holding it wrong?

**Comment by deetergp** [comment]
Disregard my prior comment. After logging out and back in again, it worked as expected 👍

**Review by deetergp — DISMISSED** [review]
👍 I'll merge once @robertKozik gives a 👍 too.

File: src/components/SpacerView.js (hunk 3 above)
Comment (robertKozik) [reply src/components/SpacerView.js]: Can you also remove the redundant eslint disable comment? It's the last thing bothering me :P

Comment (zukilover) [reply src/components/SpacerView.js]: @robertKozik done :)

**Review by robertKozik — APPROVED** [review]
Looks good 🚀 Let's push this forward

**Review by deetergp — APPROVED** [review]

**Pull request merged**